sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/DefaultStep.java

```diff
@@ -160,12 +160,15 @@ public Status getStatus() {
     public synchronized void update(Protos.TaskStatus status) {
         logger.info("Step {} received status: {}", getName(), TextFormat.shortDebugString(status));
 
-        if (!tasks.containsKey(status.getTaskId().getValue())) {
+        Optional<Protos.TaskInfo> taskInfoOptional = getTaskInfo(status.getTaskId());
```

Comment:
Maybe the work done in `getTaskInfo` could be avoided if we kept the TaskID string as the map key and then kept a (TaskInfo, Status) pair as the value? I think the semantics are clearer this way and it's a bit of computation savings. Alternatively, you could keep two side-by-side maps if you want to avoid defining an ad hoc value class. (If only there were product types in this damn language...)